tests/integration/test_generate_oci_image.py

```diff
@@ -0,0 +1,643 @@
+    _write_base_images_digests(data.base_images_path, grandparent_img)
+
+    return data
+
+
+def _get_component_gdata(
+    tmp_path: Path,
+    parent_img: Image,
+    common_packages: list[Package],
+    parent_packages: list[Package],
+    component_packages: list[Package],
+) -> GenerateData:
+    data = GenerateData(
+        input_sbom_path=tmp_path / "component.input.spdx.json",
+        input_sbom=_get_component_input_sbom(),
+        output_sbom_path=tmp_path / "component.spdx.json",
+        df_json_path=tmp_path / "component_df.json",
+        base_images_path=tmp_path / "component_base_images_digest.txt",
+    )
+
+    _add_packages_with_relationships(data.input_sbom, common_packages)
+    _add_packages_with_relationships(data.input_sbom, parent_packages)
+    _add_packages_with_relationships(data.input_sbom, component_packages)
+
+    write_file(data.input_sbom.document, str(data.input_sbom_path))
+
+    assert data.df_json_path
+    _write_parsed_dockerfile_json(data.df_json_path, parent_img)
+
+    assert data.base_images_path
+    _write_base_images_digests(data.base_images_path, parent_img)
+
+    return data
+
+
+def _write_parsed_dockerfile_json(path: Path, img: Image) -> None:
+    """
+    Write a parsed dockerfile json to the specified path, replicating a
+    DESCENDANT_OF relationship to the passed image.
+    """
+    parsed_df = {
+        "Stages": [
+            {
+                "BaseName": f"{img.repository}:{img.tag}",
+                "From": {"Image": f"{img.repository}:{img.tag}"},
+            },
+        ]
+    }
+    with open(path, "w") as fp:
+        fp.write(json.dumps(parsed_df))
+
+
+def _write_base_images_digests(path: Path, img: Image) -> None:
+    """
+    Write a base image digests file to the specified path, replicating a
+    DESCENDANT_OF relationship to the passed image.
+    """
+    with open(path, "w") as fp:
+        fp.write(
+            f"{img.repository}:{img.tag} {img.repository}:{img.tag}@{img.digest}\n"
+        )
+
+
+def _get_dependency_chain_spdx_ids(relationships: list[Relationship]) -> list[str]:
+    """
+    Using the provided relationships, builds a dependency chain based on the
+    DESCENDANT_OF relationships. The root element is the first element in the
+    list. Assumes there is only a single root.
+    """
+    rels = [
+        rel
+        for rel in relationships
+        if rel.relationship_type == RelationshipType.DESCENDANT_OF
+    ]
+
+    parents, children = set(), set()
+    for rel in rels:
+        parents.add(rel.spdx_element_id)
+        children.add(str(rel.related_spdx_element_id))
+
+    # the root element is the one that is never a child in any relationship
+    no_children = parents - children
+    root = no_children.pop()
+    assert len(no_children) == 0, "Found multiple roots in SBOM relationships."
+
+    parent_to_child = {}
+    for rel in rels:
+        parent_to_child[rel.spdx_element_id] = str(rel.related_spdx_element_id)
```

Comment:
We also could add a validation that we don't overwrite anything here (that could happen if the chain was a tree and an element would be marked as a descendant of 2 different elements).